File: runtime/src/test/java/com/siimkinks/sqlitemagic/UpdateComplexNumericColumnTest.java
package com.siimkinks.sqlitemagic;

import org.junit.Test;

import static com.google.common.truth.Truth.assertThat;
import static com.siimkinks.sqlitemagic.ComplexColumnTestFixtures.ACCOUNT;
import static com.siimkinks.sqlitemagic.ComplexColumnTestFixtures.ARTICLE;
import static com.siimkinks.sqlitemagic.ComplexColumnTestFixtures.AccountId;
import static com.siimkinks.sqlitemagic.ComplexColumnTestFixtures.Article;
import static com.siimkinks.sqlitemagic.Utils.STRING_PARSER;

public final class UpdateComplexNumericColumnTest {
  private static final Column<String, String, CharSequence, Article, NotNullable> TITLE = new Column<>(ARTICLE, "title", false, STRING_PARSER, false, null);

  @Test
  public void initialRelationshipAssignmentUsesDeclaredIdTypeAndTransformer() {
    final Update.Set<Article> update = Update
        .table(ARTICLE)
        .set(ACCOUNT, new AccountId(41L));

    assertUpdate(update, "UPDATE article SET account=? ", "42");
  }

  @Test
  public void chainedRelationshipAssignmentUsesDeclaredIdTypeAndTransformer() {
    final Update.Set<Article> update = Update
        .table(ARTICLE)
        .set(TITLE, "Title")
        .set(ACCOUNT, new AccountId(41L));

    assertUpdate(update, "UPDATE article SET title=?,account=? ", "Title", "42");
  }

  private static void assertUpdate(
      Update.Set<Article> update,
      String expectedSql,
      String... expectedArgs
  ) {
    assertThat(SqlCreator.getSql(update, update.updateBuilder.sqlNodeCount)).isEqualTo(expectedSql);
    assertThat(update.updateBuilder.args)
        .containsExactlyElementsIn(expectedArgs)
        .inOrder();
  }
}

File: runtime/src/test/kotlin/com/siimkinks/sqlitemagic/UpdateComplexNumericColumnTest.kt
package com.siimkinks.sqlitemagic

import com.google.common.truth.Truth.assertThat
import com.siimkinks.sqlitemagic.ComplexColumnTestFixtures.ACCOUNT
import com.siimkinks.sqlitemagic.ComplexColumnTestFixtures.ARTICLE
import com.siimkinks.sqlitemagic.ComplexColumnTestFixtures.AccountId
import com.siimkinks.sqlitemagic.ComplexColumnTestFixtures.Article
import com.siimkinks.sqlitemagic.Utils.STRING_PARSER
import org.junit.Test

class UpdateComplexNumericColumnTest {
  private val title: Column<String, String, CharSequence, Article, NotNullable> = Column(
    table = ARTICLE,
    name = "title",
    valueParser = STRING_PARSER,
    nullable = false,
    alias = null
  )

  @Test
  fun `initial relationship assignment uses declared ID type and transformer`() {
    val update = Update
      .table(ARTICLE)
      .set(ACCOUNT, AccountId(41L))

    assertUpdate(
      update = update,
      expectedSql = "UPDATE article SET account=? ",
      "42"
    )
  }

  @Test
  fun `chained relationship assignment uses declared ID type and transformer`() {
    val update = Update
      .table(ARTICLE)
      .set(title, "Title")
      .set(ACCOUNT, AccountId(41L))

    assertUpdate(
      update = update,
      expectedSql = "UPDATE article SET title=?,account=? ",
      "Title",
      "42"
    )
  }

  @Test
  fun `typed then raw assignments preserve SQL and argument order`() {
    val update = Update
      .table(ARTICLE)
      .set(title, "Title")
      .set("account", "42")

    assertUpdate(
      update = update,
      expectedSql = "UPDATE article SET title=?,account=? ",
      "Title",
      "42"
    )
  }

  @Test
  fun `raw then typed assignments preserve SQL and argument order`() {
    val update = Update
      .table(ARTICLE)
      .set("title", "Title")
      .set(ACCOUNT, AccountId(41L))

    assertUpdate(
      update = update,
      expectedSql = "UPDATE article SET title=?,account=? ",
      "Title",
      "42"
    )
  }

  private fun assertUpdate(
    update: Update.Set<Article>,
    expectedSql: String,
    vararg expectedArgs: String
  ) {
    assertThat(SqlCreator.getSql(update, update.updateBuilder.sqlNodeCount))
      .isEqualTo(expectedSql)
    assertThat(update.updateBuilder.args)
      .containsExactlyElementsIn(expectedArgs)
      .inOrder()
  }
}
